### PR TITLE
Add torque derating on DCM

### PR DESCRIPTION
**Pull request by timbitss**
### Summary
- Limit maximum torque based on power available from BMS
- "Powertrain" efficiency set to conservative value of 0.8 currently for all speed ranges (decrease if exceeding power limit and vice-versa)

### Checklist
*Please change `[ ]` to `[x]` when you are ready.*
- [ x] I have read and followed the code conventions detailed in [README.md](../README.md) (*This will save time for both you and the reviewer!*).
- [x ] If this pull request is longer then **500** lines, I have provided *explicit* justification in the summary above explaining why I *cannot* break this up into multiple pull requests (*Small PR's are faster and less painful for everyone involved!*).

### CODE REVIEW

File: boards/DCM/Src/App/states/App_DriveState.c
Comment (Auhv) [inline boards/DCM/Src/App/states/App_DriveState.c:18]: the .speed_rpm part of the struct seems redundant here. You could do the same thing as a simple 1-D array instead of a struct.

uint8_t eff_lut_index = speed_rpm/RPM_INTERVAL; //RPM_INTERVAL is 500 in the LUT you proposed
float efficiency_estimate = EfficiencyLUT[eff_lut_index ];

This way you're not iterating through a for loop to find the right index, and you don't need to waste memory on the struct. Will become important when you're dealing with a giant 2-D LUT with interpolation

File: boards/DCM/Src/App/states/App_DriveState.c
Comment (Auhv) [inline boards/DCM/Src/App/states/App_DriveState.c:42]: Missing abs function on this value

File: boards/DCM/Src/App/states/App_DriveState.c
Comment (Eddy13013) [inline boards/DCM/Src/App/states/App_DriveState.c:52]: maybe do different efficiency estimates per motor. because efficiency depends on speed and because the two motor speeds may be different, there could be

Comment (Auhv) [reply boards/DCM/Src/App/states/App_DriveState.c:52]: Agree

Comment (Auhv) [inline boards/DCM/Src/App/states/App_DriveState.c:52]: FSM max advertised torque should be in this MIN expression

**Review by Auhv — CHANGES_REQUESTED** [review]

File: boards/DCM/Src/App/states/App_DriveState.c
Comment (pasitlaothamatas) [inline boards/DCM/Src/App/states/App_DriveState.c:18]: the efficiency estimate is not a number bewtween [0,1]. the torque requested will be bigger tahn expected

File: boards/DCM/Src/App/states/App_DriveState.c
Comment (pasitlaothamatas) [inline boards/DCM/Src/App/states/App_DriveState.c:59]: why compute the min here? shouldnt your derated max torque request already be computed in the equation above?

Comment (pasitlaothamatas) [reply boards/DCM/Src/App/states/App_DriveState.c:59]: should just be 
`0.01f * App_CanRx_FSM_PEDAL_POSITION_GetSignal_MAPPED_PEDAL_PERCENTAGE(can_rx) * max_torque_request`

File: boards/DCM/Src/App/states/App_DriveState.c
Comment (pasitlaothamatas) [inline boards/DCM/Src/App/states/App_DriveState.c:53]: dont we want to average the motor speed? we are adding them right now

Comment (pasitlaothamatas) [reply boards/DCM/Src/App/states/App_DriveState.c:53]: actually that might be dumb

File: boards/DCM/Src/App/states/App_DriveState.c
Comment (Auhv) [reply boards/DCM/Src/App/states/App_DriveState.c:59]: Yeah I think i agree - couldn't all these things be one giant MIN statement?

File: boards/DCM/Src/App/states/App_DriveState.c
Comment (pasitlaothamatas) [reply boards/DCM/Src/App/states/App_DriveState.c:53]: dividing by the sum of the motor speeds seems incorrect though

Comment (Auhv) [reply boards/DCM/Src/App/states/App_DriveState.c:53]: Adding them is correct, it will result in a torque that doesn't need to be divided by two

File: boards/DCM/Src/App/states/App_DriveState.c
Comment (pasitlaothamatas) [inline boards/DCM/Src/App/states/App_DriveState.c:9]: I beleive this macro is alrd implemented in App_SharedMacros.h

File: boards/DCM/Src/App/states/App_DriveState.c
Comment (pasitlaothamatas) [inline boards/DCM/Src/App/states/App_DriveState.c:18]: I think it would also be nice to place the lookup table in a header file

File: boards/DCM/Src/App/states/App_DriveState.c
Comment (pasitlaothamatas) [reply boards/DCM/Src/App/states/App_DriveState.c:53]: the `bms_available_power` is the combined power of the left motor and the right motor correct? how do we optimally allocate this power to either motor right now? @Auhv

Comment (Auhv) [reply boards/DCM/Src/App/states/App_DriveState.c:53]: Same torque for both @pasitlaothamatas

File: boards/DCM/Src/App/states/App_DriveState.c
Comment (Auhv) [reply boards/DCM/Src/App/states/App_DriveState.c:52]: This way we're not arbitrarily choosing the RHS motor to calculate efficiency.

File: boards/DCM/Src/App/states/App_DriveState.c
Comment (pasitlaothamatas) [reply boards/DCM/Src/App/states/App_DriveState.c:53]: ok adding would make sense then sounds good

File: boards/DCM/Src/App/states/App_DriveState.c
Comment (pasitlaothamatas) [reply boards/DCM/Src/App/states/App_DriveState.c:18]: in the case that we have time to test the efficiency a bit better we can quickly update their values in the LUT so i dont mind keeping this for now in a header file. In our implementation in the state machine we can do what you suggested

File: boards/DCM/Src/App/states/App_DriveState.c
Comment (timbitss) [reply boards/DCM/Src/App/states/App_DriveState.c:42]: Only right motor speed needs abs, but I could just do both for now

File: boards/DCM/Src/App/states/App_DriveState.c
Comment (timbitss) [reply boards/DCM/Src/App/states/App_DriveState.c:18]: resolved

File: boards/DCM/Src/App/states/App_DriveState.c
Comment (timbitss) [reply boards/DCM/Src/App/states/App_DriveState.c:59]: Resolved

File: boards/DCM/Src/App/states/App_DriveState.c
Comment (timbitss) [reply boards/DCM/Src/App/states/App_DriveState.c:52]: Sorry is this the torque plausibility that we were talking about awhile ago. I did not have a chance to implement it

Comment (timbitss) [reply boards/DCM/Src/App/states/App_DriveState.c:52]: For now, we can keep with a single efficiency for simplicity

File: boards/DCM/Src/App/states/App_DriveState.c
Comment (timbitss) [reply boards/DCM/Src/App/states/App_DriveState.c:18]: resolved

File: boards/DCM/Src/App/states/App_DriveState.c
Comment (pasitlaothamatas) [inline boards/DCM/Src/App/states/App_DriveState.c:22]: for simplicity sake we should 
1) remove this variable
2) create a #define for the efficiency estimate and place it in this file
3) remove `EFFICIENCY_LUT_RPM_INTERVAL` in the header file (dont think its used anywhere after lut_index is removed)

we can keep the struct in the header file for now

Should be ready to merge after this

File: boards/DCM/Src/App/states/App_DriveState.c
Comment (pasitlaothamatas) [inline boards/DCM/Src/App/states/App_DriveState.c:10]: if some of these values are shared with the wheel speed sensor i suggest placing this macro in App_SharedMacros, and placing some #defines in App_SharedConstants

Comment (pasitlaothamatas) [reply boards/DCM/Src/App/states/App_DriveState.c:10]: M_PI shouldn't require a cast btw. There should be a floating pt version for PI if this M_PI is a double

File: boards/DCM/Src/App/states/App_DriveState.c
Comment (pasitlaothamatas) [inline boards/DCM/Src/App/states/App_DriveState.c:30]: Kinda nit but can we create a new float variable to hold this value and pass it later into the min macro? Will be a bit nicer to read

File: boards/DCM/Src/App/states/App_DriveState.c
Comment (timbitss) [reply boards/DCM/Src/App/states/App_DriveState.c:10]: I don't see a floating point PI

File: boards/DCM/Src/App/states/App_DriveState.c
Comment (timbitss) [reply boards/DCM/Src/App/states/App_DriveState.c]: left `EFFICIENCY_LUT_RPM_INTERVAL` incase we use it

File: boards/DCM/Src/App/states/App_DriveState.c
Comment (timbitss) [reply boards/DCM/Src/App/states/App_DriveState.c]: resolved

**Review by pasitlaothamatas — APPROVED** [review]

File: boards/DCM/Src/App/states/App_DriveState.c
Comment (Auhv) [reply boards/DCM/Src/App/states/App_DriveState.c]: Maybe now is a good time to finish that PR

**Pull request merged**
